File: problem_064.c
// -*- compile-command: "gcc -o problem_064 problem_064.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <math.h>

int main()
{
    printf("Project Euler - Problem 64:\n"
           "How many continued fractions for N <= 10000 have an odd period?\n\n");

    int number = 0, count = 0;
    while (++number < 10000)
    {
        int a0 = sqrt(number);
        if (a0 * a0 == number) continue;

        int period = 0;
        int d = 1;
        int m = 0;
        int a = a0;

        while ( a != a0 * 2)
        {
            m = d*a - m;
            d = (number - m * m) / d;
            a = (a0 + m) /d;
            period++;
        }


        if (period %2 != 0) count++;
    }

    printf("Fractions with odd period: %d\n", count);
    return 0;
}

File: problem_064.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "xcrun -sdk macosx swiftc problem_064.swift -o problem_064_swift" -*-
// Copyright (c) 2016 Michael Caldwell
import Foundation

print("Project Euler - Problem 64:")
print("How many continued fractions for N <= 10000 have an odd period?\n")

var number = 0
var count = 0
let debug = false

while number < 10000 {
    number += 1
    let a0 = Int(sqrt(Float(number)))

    // Check if perfect square; skip
    if a0 * a0 == number {
        continue
    }

    var period = 0
    var d = 1
    var m = 0
    var a = a0

    while a != a0 * 2 {
        m = d * a - m
        d = (number - (m * m)) / d
        a = (a0 + m) / d
        period += 1
    }


    if period % 2 != 0 {
        count += 1
    }

    if debug == true {
        print(number, period)
    }
}

print(count)
